File: src/novel_generator/phase0/seed_generator.py
"""Seed generator using LLM to create NovelSeed from interview results."""

import re

import yaml

from novel_generator.schema.novel import NovelSeed
from novel_generator.phase0.prompts import SEED_GENERATION_PROMPT


class SeedGenerator:
    """Generates NovelSeed from interview results.

    Uses ouroboros LLM client for generation.
    """

    def __init__(self, use_mock: bool = False):
        """Initialize generator.

        Args:
            use_mock: If True, use mock response instead of LLM (for testing)
        """
        self.use_mock = use_mock

    async def generate(self, interview_result: str) -> NovelSeed:
        """Generate NovelSeed from interview result.

        Args:
            interview_result: Summary from NovelInterviewer

        Returns:
            Generated and validated NovelSeed
        """
        prompt = SEED_GENERATION_PROMPT.format(interview_result=interview_result)

        if self.use_mock:
            response = self._mock_response(interview_result)
        else:
            response = await self._call_llm(prompt)

        # Parse YAML from response
        seed_data = self._parse_yaml_response(response)

        # Validate and create NovelSeed
        return NovelSeed.model_validate(seed_data)

    async def _call_llm(self, prompt: str) -> str:
        """Call LLM to generate seed via ouroboros adapter."""
        from novel_generator.ouroboros.llm import call_llm

        return await call_llm(
            prompt=prompt,
            system="당신은 한국 웹소설 기획 전문가입니다. YAML 형식으로 출력하세요.",
            temperature=0.7,
            max_tokens=8000,
        )

    def _parse_yaml_response(self, response: str) -> dict:
        """Extract and parse YAML from LLM response."""
        # Try to find YAML block
        yaml_match = re.search(r'```ya?ml\s*(.*?)```', response, re.DOTALL)
        if yaml_match:
            yaml_str = yaml_match.group(1)
        else:
            # Try to parse entire response as YAML
            yaml_str = response

        # Fix common YAML issues from LLM
        yaml_str = self._fix_yaml(yaml_str)

        try:
            return yaml.safe_load(yaml_str)
        except yaml.YAMLError as e:
            raise ValueError(f"Failed to parse YAML: {e}")

    def _fix_yaml(self, yaml_str: str) -> str:
        """Fix common YAML formatting issues from LLM output."""
        lines = yaml_str.split('\n')
        fixed_lines = []

        for line in lines:
            # Fix inline arrays like: - "a", "b", "c" -> proper YAML list
            if re.match(r'^(\s*)-\s*"[^"]+",\s*"', line):
                indent = len(line) - len(line.lstrip())
                # Extract all quoted strings
                items = re.findall(r'"([^"]*)"', line)
                for item in items:
                    fixed_lines.append(' ' * indent + f'- "{item}"')
            # Fix inline arrays like: - ~하냐?, ~잖아! (without quotes)
            elif re.match(r'^(\s*)-\s*~[^,]+,', line):
                indent = len(line) - len(line.lstrip())
                # Split by comma and create separate items
                content = line.lstrip('- ').strip()
                items = [item.strip() for item in content.split(',')]
                for item in items:
                    fixed_lines.append(' ' * indent + f'- "{item}"')
            # Fix trailing comments/annotations like: - "값" (설명) -> - "값"
            elif re.match(r'^(\s*)-\s*"[^"]+"\s*\(', line):
                indent = len(line) - len(line.lstrip())
                # Extract just the quoted value
                match = re.search(r'"([^"]*)"', line)
                if match:
                    fixed_lines.append(' ' * indent + f'- "{match.group(1)}"')
                else:
                    fixed_lines.append(line)
            # Fix unquoted values with trailing comments: - 값 (설명) -> - "값"
            elif re.match(r'^(\s*)-\s*[^"\[\{]+\s*\(', line):
                indent = len(line) - len(line.lstrip())
                # Extract value before parenthesis
                match = re.match(r'^(\s*)-\s*([^(]+)\s*\(', line)
                if match:
                    value = match.group(2).strip()
                    fixed_lines.append(' ' * indent + f'- "{value}"')
                else:
                    fixed_lines.append(line)
            else:
                fixed_lines.append(line)

        return '\n'.join(fixed_lines)

    def _mock_response(self, interview_result: str) -> str:
        """Generate mock response for testing without LLM."""
        # Extract some info from interview for mock
        genre = "현대 판타지"
        sub_genre = "회귀"

        if "무협" in interview_result:
            genre = "무협"
            sub_genre = "회귀"
        elif "로맨스" in interview_result or "로판" in interview_result:
            genre = "로맨스"
            sub_genre = "빙의"

        return f'''```yaml
title: "회귀한 천재의 두 번째 삶"
logline: "죽음 직전 과거로 돌아온 천재가 모든 것을 바꾸려 한다"
total_chapters: 300

world:
  name: "현대 한국 + 헌터 세계관"
  genre: "{genre}"
  sub_genre: "{sub_genre}"
  time_period: "현대"
  magic_system: "마나를 이용한 스킬 시스템. 등급은 F~S"
  key_locations:
    academy: "한국 최고의 헌터 양성 아카데미"
    gate: "몬스터가 출현하는 차원문"
    guild: "상위 길드 본부"
  factions:
    hunter_association: "헌터들을 관리하는 정부 기관"
    top_guilds: "S급 헌터들이 이끄는 대형 길드들"
    villains: "게이트를 이용하는 범죄 조직"
  rules:
    - "각성자만이 마나를 다룰 수 있다"
    - "등급은 각성 시 결정되며 변하지 않는다 (주인공 제외)"
    - "S급은 전 세계에 100명 미만"

characters:
  - id: "mc"
    name: "강현우"
    role: "주인공"
    introduction_chapter: 1
    voice:
      tone: "냉소적이지만 속정 있음. 회귀 전 트라우마로 인한 경계심"
      speech_patterns:
        - "~하지"
        - "...그래서?"
        - "알아서 해"
      sample_dialogues:
        - "또 이 꼴이군. 지겹다, 진짜."
        - "...네가 뭘 알아."
        - "이번엔 다르게 할 거야. 반드시."
        - "걱정 마. 죽지 않아."
      personality_core: "과거의 실패로 냉소적이 되었으나, 이번 생에선 소중한 사람들을 지키겠다는 강한 의지"
    backstory: "전생에서 S급까지 올랐으나 모든 것을 잃고 죽음. 10년 전으로 회귀"
    arc_summary: "냉소적 복수자 → 동료를 얻으며 변화 → 진정한 영웅으로 성장"
    state:
      level: 1
      relationships: {{}}

  - id: "heroine"
    name: "이서연"
    role: "히로인"
    introduction_chapter: 3
    voice:
      tone: "밝고 당찬, 하지만 자신만의 상처가 있음"
      speech_patterns:
        - "~거든요!"
        - "에이, 설마~"
      sample_dialogues:
        - "포기하면 거기서 끝이에요!"
        - "...저도 지키고 싶은 사람이 있어요."
      personality_core: "겉으로는 밝지만 가족을 잃은 트라우마를 숨기고 있음"
    backstory: "전생에서 주인공이 지키지 못한 사람. 이번엔 반드시 지킨다"
    arc_summary: "밝은 동료 → 주인공의 비밀을 알게 됨 → 함께 싸우는 파트너"
    state:
      level: 1
      relationships: {{}}

arcs:
  - id: "arc_1"
    name: "귀환편"
    start_chapter: 1
    end_chapter: 50
    summary: "회귀 후 적응하며 첫 번째 위기를 막는다"
    key_events:
      - "회귀 각성"
      - "아카데미 입학"
      - "히로인과 만남"
      - "첫 번째 대형 게이트 사건"
      - "숨겨진 던전 발견"
    climax_chapter: 48

  - id: "arc_2"
    name: "각성편"
    start_chapter: 51
    end_chapter: 120
    summary: "S급으로의 첫 걸음, 길드 설립"
    key_events:
      - "등급 돌파의 비밀"
      - "첫 S급 헌터와 대결"
      - "길드 설립"
    climax_chapter: 118

chapter_outlines:
  - chapter_number: 1
    title: "다시, 처음으로"
    arc_id: "arc_1"
    one_liner: "죽음의 순간, 10년 전으로 돌아온 강현우"
    key_points:
      - "전생의 마지막 전투 회상"
      - "눈을 떠보니 10년 전 자신의 방"
      - "회귀 확신"
    characters_involved:
      - "mc"
    tension_level: 8

  - chapter_number: 2
    title: "달라진 것과 달라지지 않은 것"
    arc_id: "arc_1"
    one_liner: "과거의 기억을 확인하며 계획을 세우는 현우"
    key_points:
      - "현재 상황 파악"
      - "바꿔야 할 것들 정리"
      - "첫 번째 목표 설정"
    characters_involved:
      - "mc"
    tension_level: 4

  - chapter_number: 3
    title: "우연이 아닌 만남"
    arc_id: "arc_1"
    one_liner: "서연과의 첫 만남, 이번엔 다르게"
    key_points:
      - "히로인 등장"
      - "전생과 다른 첫인상"
      - "복선: 히로인의 과거"
    characters_involved:
      - "mc"
      - "heroine"
    tension_level: 5

foreshadowing:
  - id: "fs_ring"
    name: "검은 반지의 비밀"
    description: "주인공이 가진 검은 반지의 정체"
    importance: "critical"
    planted_at: 5
    hints_at: [15, 30, 42]
    reveal_at: 48

  - id: "fs_heroine_past"
    name: "히로인의 과거"
    description: "히로인 가족의 죽음과 관련된 진실"
    importance: "critical"
    planted_at: 3
    hints_at: [20, 35]
    reveal_at: 45

  - id: "fs_regression"
    name: "회귀의 대가"
    description: "회귀에는 아직 밝혀지지 않은 대가가 있다"
    importance: "normal"
    planted_at: 1
    hints_at: [25, 50]
    reveal_at: 120

style:
  max_paragraph_length: 3
  dialogue_ratio: 0.6
  sentence_style: "short"
  hook_ending: true
  pov: "1인칭"
  tense: "과거형"
  formatting_rules:
    - "문단은 3문장 이하로"
    - "대사 후 긴 지문 금지"
    - "클리셰 표현 사용 가능 (장르 특성)"
    - "매 회차 끝은 궁금증 유발"
```'''

    def validate_seed(self, seed: NovelSeed) -> list[str]:
        """Validate seed for completeness and consistency.

        Returns:
            List of validation warnings (empty if valid)
        """
        warnings = []

        # Check characters
        if len(seed.characters) < 2:
            warnings.append("캐릭터가 2명 미만입니다. 최소 주인공과 히로인/조력자 필요")

        for char in seed.characters:
            if len(char.voice.sample_dialogues) < 3:
                warnings.append(f"캐릭터 '{char.name}'의 대사 예시가 부족합니다 (최소 3개)")

        # Check arcs
        if not seed.arcs:
            warnings.append("스토리 아크가 정의되지 않았습니다")

        # Check arc coverage
        if seed.arcs:
            covered = set()
            for arc in seed.arcs:
                covered.update(range(arc.start_chapter, arc.end_chapter + 1))

            total_range = set(range(1, seed.total_chapters + 1))
            uncovered = total_range - covered
            if uncovered:
                first_uncovered = min(uncovered)
                warnings.append(f"{first_uncovered}화부터 아크가 정의되지 않았습니다")

        # Check foreshadowing
        if len(seed.foreshadowing) < 3:
            warnings.append("복선이 3개 미만입니다. 장편에선 더 많은 복선 권장")

        critical_fs = [fs for fs in seed.foreshadowing if fs.importance == "critical"]
        if not critical_fs:
            warnings.append("중요 복선(critical)이 없습니다")

        # Check chapter outlines
        if len(seed.chapter_outlines) < 10:
            warnings.append("챕터 아웃라인이 10개 미만입니다")

        return warnings

File: web/src/lib/evolution/seed-generator.ts
import { z } from "zod";
import { getAgent } from "@/lib/agents/llm-agent";
import type { TokenUsage } from "@/lib/agents/types";
import { runPlotPipeline } from "@/lib/agents/plot-pipeline";
import type { NovelSeed } from "@/lib/schema/novel";
import {
  WorldSettingSchema,
  PlotArcSchema,
  ChapterOutlineSchema,
  StyleGuideSchema,
} from "@/lib/schema/novel";
import { CharacterSchema } from "@/lib/schema/character";
import { ForeshadowingSchema } from "@/lib/schema/foreshadowing";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface SeedGeneratorConfig {
  genre: string;
  /** Optional premise/logline to use (if not provided, generates one via plot pipeline) */
  premise?: string;
  /** Target total chapters (default: 200) */
  totalChapters?: number;
  /** Model to use for generation */
  model?: string;
  /** Progress callback */
  onProgress?: (step: string) => void;
}

export interface SeedGeneratorResult {
  seed: NovelSeed;
  usage: TokenUsage;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function emptyUsage(): TokenUsage {
  return { prompt_tokens: 0, completion_tokens: 0, total_tokens: 0, cost_usd: 0 };
}

function addUsage(a: TokenUsage, b: TokenUsage): TokenUsage {
  return {
    prompt_tokens: a.prompt_tokens + b.prompt_tokens,
    completion_tokens: a.completion_tokens + b.completion_tokens,
    total_tokens: a.total_tokens + b.total_tokens,
    cost_usd: a.cost_usd + b.cost_usd,
  };
}

// ---------------------------------------------------------------------------
// Inline Zod schemas for each step
// ---------------------------------------------------------------------------

const WorldStepSchema = WorldSettingSchema;

const CharacterStepSchema = z.array(CharacterSchema);

const StoryStructureSchema = z.object({
  arcs: z.array(PlotArcSchema),
  chapter_outlines: z.array(ChapterOutlineSchema),
});

const ForeshadowingStepSchema = z.array(ForeshadowingSchema);

// ---------------------------------------------------------------------------
// Main pipeline
// ---------------------------------------------------------------------------

export async function generateRichSeed(
  config: SeedGeneratorConfig,
): Promise<SeedGeneratorResult> {
  const {
    genre,
    totalChapters = 200,
    model,
    onProgress,
  } = config;

  const agent = getAgent();
  let usage = emptyUsage();
  const callOpts = model ? { model } : {};

  // =========================================================================
  // Step 1: Premise
  // =========================================================================
  onProgress?.("Step 1/5: 프리미스 생성 중...");

  let title: string;
  let logline: string;
  let arcSummary: string[];

  if (config.premise) {
    title = "";
    logline = config.premise;
    arcSummary = [];

    // Generate a title from the premise
    const titleResult = await agent.call({
      prompt: `다음 프리미스로 ${genre} 웹소설 제목을 하나만 생성하세요. 제목만 출력하세요.\n\n프리미스: ${config.premise}`,
      system: "당신은 한국 웹소설 제목 전문가입니다.",
      temperature: 0.7,
      maxTokens: 100,
      taskId: "seed-title",
      ...callOpts,
    });
    title = titleResult.data.trim().replace(/^["'「『]|["'」』]$/g, "");
    usage = addUsage(usage, titleResult.usage);
  } else {
    const plotResult = await runPlotPipeline(genre);
    usage = addUsage(usage, plotResult.usage);

    const best = plotResult.plots[0];
    if (!best) throw new Error("Plot pipeline returned no plots");

    title = best.title;
    logline = best.logline;
    arcSummary = best.arc_summary;
  }

  // =========================================================================
  // Step 2: World Building
  // =========================================================================
  onProgress?.("Step 2/5: 세계관 설계 중...");

  const worldPrompt = `다음 프리미스를 가진 ${genre} 웹소설의 세계관을 설계하세요.

제목: ${title}
로그라인: ${logline}
${arcSummary.length > 0 ? `아크 요약:\n${arcSummary.map((a, i) => `${i + 1}. ${a}`).join("\n")}` : ""}

다음 요소를 반드시 포함하세요:
- name: 세계/배경의 이름
- genre: "${genre}"
- sub_genre: 구체적인 서브장르
- time_period: 시대적 배경
- magic_system: 마법/능력 체계 (해당 없으면 null)
- key_locations: 최소 4개 이상의 주요 장소 (장소명: 설명)
- factions: 최소 3개 이상의 세력/조직 (이름: 설명)
- rules: 최소 3개 이상의 세계 규칙

JSON 형식으로 출력하세요.`;

  const worldResult = await agent.callStructured({
    prompt: worldPrompt,
    system: "당신은 한국 웹소설 세계관 설계 전문가입니다.",
    temperature: 0.7,
    maxTokens: 4096,
    schema: WorldStepSchema,
    format: "json",
    taskId: "seed-world",
    ...callOpts,
  });
  const world = worldResult.data;
  usage = addUsage(usage, worldResult.usage);

  // =========================================================================
  // Step 3: Character Sheet
  // =========================================================================
  onProgress?.("Step 3/5: 캐릭터 설계 중...");

  const worldContext = JSON.stringify(world, null, 2);

  const characterPrompt = `다음 세계관과 프리미스를 바탕으로 8~12명의 캐릭터를 설계하세요.

제목: ${title}
로그라인: ${logline}

세계관:
${worldContext}

각 캐릭터에 대해 다음을 포함하세요:
- id: 고유 ID (예: "char_1", "char_2")
- name: 한국식 이름
- role: 역할 (주인공, 히로인, 악역, 조력자, 멘토 등)
- introduction_chapter: 첫 등장 화수 (주인공은 1, 나머지는 1~30 사이에 분산)
- voice: 캐릭터 목소리 설정
  - tone: 전체적인 말투 톤
  - speech_patterns: 최소 3개 이상의 고유한 말투 패턴
  - sample_dialogues: 최소 5개 이상의 대표 대사 (실제 소설에서 사용할 수 있는 수준)
  - personality_core: 핵심 성격 설명
- backstory: 배경 스토리
- arc_summary: 캐릭터 성장 아크 요약
- state: 초기 상태
  - location: 시작 위치
  - status: "normal"
  - relationships: 다른 캐릭터와의 관계 (이름: 관계)

**핵심 요구사항**: 각 캐릭터의 대사 샘플은 즉시 구분 가능해야 합니다. 두 캐릭터가 같은 존댓말/반말 수준이나 말투 패턴을 공유해서는 안 됩니다.
예를 들어:
- 주인공은 짧고 무뚝뚝한 반말 ("...알았어.", "그래서?")
- 히로인은 활발한 반말 ("진짜?! 대박이다!", "야야야 잠깐만!")
- 멘토는 권위 있는 존댓말 ("그것은 네가 결정할 일이다.", "배움에는 끝이 없느니라.")

JSON 배열 형식으로 출력하세요.`;

  const characterResult = await agent.callStructured({
    prompt: characterPrompt,
    system: "당신은 캐릭터 설계 전문가입니다. 각 캐릭터의 목소리가 확실히 구분되어야 합니다.",
    temperature: 0.7,
    maxTokens: 8192,
    schema: CharacterStepSchema,
    format: "json",
    taskId: "seed-characters",
    ...callOpts,
  });
  const characters = characterResult.data;
  usage = addUsage(usage, characterResult.usage);

  // =========================================================================
  // Step 4: Story Structure (Arcs + Chapter Outlines)
  // =========================================================================
  onProgress?.("Step 4/5: 스토리 구조 설계 중...");

  const characterSummary = characters
    .map((c) => `- ${c.name} (${c.role}): ${c.arc_summary}`)
    .join("\n");

  const structurePrompt = `다음 세계관과 캐릭터를 바탕으로 ${totalChapters}화 장편 웹소설의 스토리 구조를 설계하세요.

제목: ${title}
로그라인: ${logline}

세계관: ${world.name} (${world.genre}/${world.sub_genre})
주요 장소: ${Object.keys(world.key_locations).join(", ")}
세력: ${Object.keys(world.factions).join(", ")}

캐릭터:
${characterSummary}

다음을 설계하세요:

1. arcs (플롯 아크) - 처음 60화를 **6~8개 아크**로 분할:
   - ⚠️ 각 아크는 반드시 **8~15화** 길이여야 합니다. 20화 이상의 아크는 금지!
   - 예시: 아크1(1~10화), 아크2(11~20화), 아크3(21~28화), 아크4(29~38화), ...
   - 각 아크: id, name, start_chapter, end_chapter, summary, key_events (3~5개), climax_chapter, theme, tension_curve
   - theme: 아크의 주제적 핵심 (예: "신뢰와 배신", "각성과 대가")
   - tension_curve: 아크 내 각 화의 긴장도 수치 배열 (1~10). 배열 길이 = end_chapter - start_chapter + 1
   - 아크 간 전환이 자연스러워야 합니다 (이전 아크의 결말이 다음 아크의 시작을 유발)

2. chapter_outlines (회차 개요) - 처음 10화:
   - 각 회차: chapter_number, title, arc_id, one_liner, key_points (3개 이상), characters_involved, tension_level (1~10)
   - 1화는 캐릭터 2명 이하만 등장. 이후 화마다 1명씩 추가
   - tension_level은 1화에서 시작해 점진적으로 상승해야 합니다

**아크 크기 규칙 (절대 준수)**:
- 최소 8화, 최대 15화
- 60화에 아크 1~2개만 있으면 틀린 것입니다
- 카카오페이지 인기작처럼: 도입(10화) → 첫 위기(8화) → 성장(12화) → 대결(10화) → 전환(8화) → ...

JSON 형식으로 다음 구조로 출력하세요:
{
  "arcs": [...],
  "chapter_outlines": [...]
}`;

  const structureResult = await agent.callStructured({
    prompt: structurePrompt,
    system: "당신은 200화 장편 웹소설의 구조를 설계하는 전문가입니다.",
    temperature: 0.5,
    maxTokens: 8192,
    schema: StoryStructureSchema,
    format: "json",
    taskId: "seed-structure",
    ...callOpts,
  });
  const { arcs, chapter_outlines } = structureResult.data;
  usage = addUsage(usage, structureResult.usage);

  // =========================================================================
  // Step 5: Foreshadowing
  // =========================================================================
  onProgress?.("Step 5/5: 복선 설계 중...");

  const arcSummaryForForeshadowing = arcs
    .map((a) => `- ${a.name} (${a.start_chapter}~${a.end_chapter}화): ${a.summary}`)
    .join("\n");

  const foreshadowingPrompt = `다음 스토리 구조를 바탕으로 5~8개의 복선(foreshadowing)을 설계하세요.

제목: ${title}
로그라인: ${logline}

아크 구조:
${arcSummaryForForeshadowing}

캐릭터:
${characterSummary}

각 복선에 대해:
- id: 고유 ID (예: "fs_1", "fs_2")
- name: 복선의 짧은 이름
- description: 복선의 구체적 내용
- importance: "critical" 또는 "normal" 또는 "minor"
- planted_at: 복선을 심는 화수
- hints_at: 힌트를 주는 화수 배열 (3~5개, 5~8화 간격으로 분산)
- reveal_at: 복선을 회수하는 화수
- status: "pending"
- hint_count: 0

**복선 설계 규칙 (절대 준수)**:
1. 심기(planted_at)와 회수(reveal_at) 사이에 최소 3개 힌트가 있어야 합니다
2. 힌트 간격은 5~8화입니다 (2화에 심고 48화에 회수하면서 힌트 없이 방치하는 것은 금지!)
3. 회수(reveal_at)가 전부 같은 화에 몰리면 안 됩니다. 복선마다 다른 아크에서 회수
4. 복선은 반드시 아크를 가로질러야 합니다 (같은 아크 안에서 심고 회수하는 건 복선이 아님)
5. 최소 2개는 importance가 "critical"이어야 합니다

**나쁜 예 (금지)**:
  fs_1: planted_at=2, hints_at=[10], reveal_at=48  → 힌트 1개, 간격 38화 = 독자가 까먹음
  fs_2: planted_at=4, hints_at=[], reveal_at=48    → 힌트 0개, 회수가 fs_1과 같은 화 = 금지

**좋은 예**:
  fs_1: planted_at=2, hints_at=[8, 14, 20], reveal_at=25     → 힌트 3개, 균일 간격
  fs_2: planted_at=5, hints_at=[12, 19, 27, 34], reveal_at=38 → 힌트 4개, 다른 아크에서 회수

JSON 배열 형식으로 출력하세요.`;

  const foreshadowingResult = await agent.callStructured({
    prompt: foreshadowingPrompt,
    system: "당신은 복선 설계 전문가입니다.",
    temperature: 0.5,
    maxTokens: 4096,
    schema: ForeshadowingStepSchema,
    format: "json",
    taskId: "seed-foreshadowing",
    ...callOpts,
  });
  const foreshadowing = foreshadowingResult.data;
  usage = addUsage(usage, foreshadowingResult.usage);

  // =========================================================================
  // Assemble the seed
  // =========================================================================
  onProgress?.("시드 조립 완료!");

  const seed: NovelSeed = {
    title,
    logline,
    total_chapters: totalChapters,
    world,
    characters,
    arcs,
    chapter_outlines,
    foreshadowing,
    style: StyleGuideSchema.parse({}),
  };

  return { seed, usage };
}
